src/sync.rs

```diff
@@ -3,6 +3,46 @@ use std::convert::TryFrom;
 use std::io::Cursor;
 use std::prelude::v1::*;
 
+struct Buffer<R: std::io::Read> {
+    reader: R,
+    buf: Box<[u8]>,
+    level: usize,
+}
+
+impl<R: std::io::Read> Buffer<R> {
+    fn new(size: usize, reader: R) -> Self {
+        Self {
+            reader,
+            buf: vec![0; size].into_boxed_slice(),
+            level: 0,
+        }
+    }
+
+    fn fill_buf(&mut self) -> Result<&[u8], std::io::Error> {
+        while self.level < self.buf.len() {
+            let dst = &mut self.buf[self.level..];
+            let r = self.reader.read(dst)?;
+            if r == 0 {
+                break;
+            } else {
+                self.level += r;
+            }
+        }
+        Ok(&self.buf[0..self.level])
+    }
+
+    fn consume(&mut self, amt: usize) {
+        match self.level.cmp(&amt) {
+            core::cmp::Ordering::Equal => self.level = 0,
+            core::cmp::Ordering::Greater => {
+                self.buf.copy_within(amt..self.level, 0);
+                self.level -= amt
+            }
+            core::cmp::Ordering::Less => panic!("Tried to consume more then what was left"),
```

Comment:
I'm not sure if we should panic here https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4c46783b3d35819204994c2b5e67edaf

The [doc on BufRead](https://doc.rust-lang.org/std/io/trait.BufRead.html#tymethod.consume) is not really specific about what should be the behavior but I think it's best to stay close to what is the current known behavior.

Reply:
I assumed htat must basically means; if it's not it might panic :) but reading the BufReader source it indeed clamps so doing the same now :)

Reply:
Yes the wording isn't super specific. If it was really a panic you would have a "Panics" title listing all the cases where it can panic. But here it sounds more like UB. Yet the function isn't unsafe. Maybe it's just because it's a trait so the implementation has some flexibility? In which case, the simpler the better I guess... that was my reasoning.

Reply:
That wasn't blocking for me to merge btw